bfsbverify

```diff
@@ -0,0 +1,1061 @@
+    # 	subject              Name,
+    # 	subjectPublicKeyInfo SubjectPublicKeyInfo,
+    # 	issuerUniqueID  [1]  IMPLICIT UniqueIdentifier OPTIONAL,
+    #   	                   -- If present, version MUST be v2 or v3
+    # 	subjectUniqueID [2]  IMPLICIT UniqueIdentifier OPTIONAL,
+    #       	               -- If present, version MUST be v2 or v3
+    #	extensions      [3]  Extensions OPTIONAL
+    #
+    #
+    off=4
+    len=$(openssl asn1parse -inform der -in $in \
+            | head -n 2 | grep "4:d=1" | cut -d"=" -f4 | tr -dc '0-9')
+    # Include header (4 bytes)
+    dd if=$in of=$out skip=$off bs=$(expr $len + 4) count=1 iflag=skip_bytes \
+        2> /dev/null
+}
+
+read_signature_value()
+{
+    in=$1
+    out=$2
+
+    #
+    # Extract the certificate signature value. The signatureValue field
+    # contains a digital signature computed upon the ASN.1 DER encoded
+    # TBSCertificate.
+    #
+    off=$(openssl asn1parse -inform der -in $in \
+            | tail -n 1 | cut -d":" -f1 | tr -dc '0-9')
+    len=512
+    # Skip header (4 bytes) and one extra byte (zero byte)
+    dd if=$in of=$out skip=$(expr $off + 4 + 1) bs=$len count=1 iflag=skip_bytes \
+        2> /dev/null
+}
+
+read_extension_value()
+{
+    oid=$1
+    in=$2
+    out=$3
+
+    # The extension value follows static bytes "3051300D060960864801650304020305000440".
+    #
+    # Here's an exemple in DER ASN.1 format:
+    #    857:d=5  hl=2 l=  12 prim: OBJECT            :1.3.6.1.4.1.33049.2100.201
+    #    871:d=5  hl=2 l=   1 prim: BOOLEAN           :255
+    #    874:d=5  hl=2 l=  83 prim: OCTET STRING      [HEX DUMP]:3051300D0609608648016503040203050004406410682AFF87ECC0B8081DCF2D5DF17DFA1978CACA259EA4C5A6EF881989E85D7B8A270DF640E05CDF95233CFF087EDBC167342F723064F1CFBF29795612905B
+    #
+    off=$(openssl asn1parse -inform der -in $in \
+            | grep "$oid" | cut -d":" -f1 | tr -dc '0-9')
+    len=64
+    # Skip the following:
+    #   OID object: header (2 bytes) and data (12 bytes),
+    #   Flag: header (2 bytes) and data (1 byte),
+    #   Value: header (2 bytes) and static bytes (19 bytes)
+    dd if=$in of=$out skip=$(expr $off + 38) bs=$len count=1 iflag=skip_bytes \
+        2> /dev/null
+}
+
+verify_cert_sig()
+{
+    cert=$1
+
+    certkey=${tmpdir}/pkey.pem
+    tbscert=${tmpdir}/tbscertificate.der
+    sigval=${tmpdir}/sigvalue.der
+
+    rm -f $certkey $tbscert $sigval
+
+    # Determine the Signature Algorithm.
+    # Only 'rsassaPss' and 'sha512WithRSAEncryption' are supported.
+    alg=$(openssl x509 -inform der -in $cert -text -noout \
+            -certopt ca_default -certopt no_validity -certopt no_serial \
+            -certopt no_subject -certopt no_extensions -certopt no_signame \
+            | grep "Signature Algorithm:" | cut -d":" -f2 | tr -d '[:space:]')
+    if [ "$alg" != "rsassaPss" ] && [ "$alg" != "sha512WithRSAEncryption" ]; then
+        echo "** Signature algorith '$alg' unsupported"
+        exit 1
```

Comment:
There are "exit" cases. We could consider using 'trap' to remove the tmp files (like bfver does... up to you).

Reply:
I have done it in purpose; I wanted to leave the temp files for debug purposes. It would be very easy to figure out the root cause of the failure.